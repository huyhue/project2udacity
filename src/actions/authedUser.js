export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_USER = "LOGOUT_USER";

export function setAuthedUser(id) {
  return {
    type: SET_AUTHED_USER,
    id,
  };
}

export function handleUserLogin(userID, pw) {
    return (dispatch, getState) => {
        const { users } = getState();

        const authenUser = Object.values(users).find((u) => 
            u.id === userID.toLowerCase() 
            && 
            u.password === pw);
        
        if(!authenUser) {
            window.alert("Wrong username or password!");
        }

        if (!!authenUser) {
            return dispatch(setAuthedUser(authenUser));
        }
    };
}

export function logoutUser() {
    return {
        type: LOGOUT_USER,
    };
}

export function handleUserLogout() {
    return (dispatch) => {
        return dispatch(logoutUser());
    }
}