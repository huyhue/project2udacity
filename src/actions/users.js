export const RECEIVE_USERS = "RECEIVE_USERS";
export const CREATE_QUESTION_USER = "CREATE_QUESTION_USER";
export const ANSWER_QUESTION_USER = "ANSWER_QUESTION_USER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function createQuestionUser({ id, author }) {
    return {
        type: CREATE_QUESTION_USER,
        id,
        author,
    }
}

export function answerQuestionUser(authedUser, qid, answer) {
    return {
        type: ANSWER_QUESTION_USER,
        authedUser,
        qid,
        answer,
    };
}