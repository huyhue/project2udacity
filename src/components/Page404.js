import { connect } from "react-redux";

const Page404 = () => {
    return(
        <div className="d-flex align-items-center justify-content-center vh-100">
            <div className="text-center">
                <h1 data-testid="notif1" className="display-1 fw-bold">404</h1>
                <p data-testid="notif2" className="fs-3"> <span className="text-danger">Opps!</span> Page not found.</p>
                <p className="lead" data-testid="notif3">
                    The page you’re looking for doesn’t exist.
                  </p>
            </div>
        </div>
    )
}
export default connect()(Page404);