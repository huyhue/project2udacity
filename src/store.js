import reducers from './reducers';
import middleware from "./middlewares";
import { legacy_createStore as createStore } from "redux";

const store = createStore(reducers, middleware);

export default store;
