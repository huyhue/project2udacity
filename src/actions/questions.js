import { createQuestionUser, answerQuestionUser } from "./users"; 
import { saveQuestion, saveQuestionAnswer } from "../utils/API";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export const CREATE_QUESTION = "CREATE_QUESTION";
export const QUESTIONS_NEED_TO_ANSWER = "QUESTIONS_NEED_TO_ANSWER";
export const CREATE_ANSWER_TO_QUESTION = "CREATE_ANSWER_TO_QUESTION";


export function needAnsweringQuestions(questions) {
    return {
        type: QUESTIONS_NEED_TO_ANSWER,
        questions,
    };
}

export function createNewQuestion(question) {
    return {
        type: CREATE_QUESTION,
        question,
    };
}

export function createAnswerToQuestion(author, qid, answer) {
    return {
        type: CREATE_ANSWER_TO_QUESTION,
        author,
        qid,
        answer,
    };
}

export function handleCreateAnswer(questionId, answer) {
    return (dispatch, getState) => {
        const { authedUser } = getState();
        return saveQuestionAnswer(authedUser.id, questionId, answer)
            .then(() => {
                dispatch(answerQuestionUser(authedUser.id, questionId, answer));
                dispatch(createAnswerToQuestion(authedUser.id, questionId, answer));
            });
    }
}

export function handleCreateQuestion(firstOption, secondOption, authedUser) {
    return (dispatch, getState) => {
        // const { authedUser } = getState();
        dispatch(showLoading());
        return saveQuestion(
            firstOption,
            secondOption,
            authedUser,
        )
          .then((question) => {
            dispatch(createNewQuestion(question));
            dispatch(createQuestionUser(question));
          })
          .then(() => dispatch(hideLoading()));
      };
}

