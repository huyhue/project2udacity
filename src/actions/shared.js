import { getInitialData } from "../utils/API";
import { needAnsweringQuestions } from "./questions"; 
import { receiveUsers } from "./users";

export function handleInitData() {
    return (dispatch) => {
        return getInitialData().then(({ users, questions }) => {
            for (const userId in users) {
                if (users.hasOwnProperty(userId)) {
                    const data = localStorage.getItem(userId);
                    if (!data) {
                        localStorage.setItem(`${userId}`, users[userId].questions.length);
                    }
                }
              }
            dispatch(receiveUsers(users));
            dispatch(needAnsweringQuestions(questions));
        })
    }
}