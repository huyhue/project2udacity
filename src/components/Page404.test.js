import { render } from "@testing-library/react";
import '@testing-library/jest-dom'
import React from "react";
import {Provider} from "react-redux";
import store from "../store";
import {BrowserRouter} from "react-router-dom";
import Page404 from "./Page404";
import {handleInitData} from "../actions/shared";

describe("Not found page", () => {
    it("should render screen", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Page404 />
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it("should show notification text", async () => {
        await store.dispatch(handleInitData());
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Page404 />
                </BrowserRouter>
            </Provider>
        );

        const notif1 = component.getByTestId("notif1");
        const notif2 = component.getByTestId("notif2");
        const notif3 = component.getByTestId("notif3");

        expect(notif1).toBeInTheDocument();
        expect(notif2).toBeInTheDocument();
        expect(notif3).toBeInTheDocument();
    });
})