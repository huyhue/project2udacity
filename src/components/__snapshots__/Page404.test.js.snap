// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not found page should render screen 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="d-flex align-items-center justify-content-center vh-100"
      >
        <div
          class="text-center"
        >
          <h1
            class="display-1 fw-bold"
            data-testid="notif1"
          >
            404
          </h1>
          <p
            class="fs-3"
            data-testid="notif2"
          >
             
            <span
              class="text-danger"
            >
              Opps!
            </span>
             Page not found.
          </p>
          <p
            class="lead"
            data-testid="notif3"
          >
            The page you’re looking for doesn’t exist.
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex align-items-center justify-content-center vh-100"
    >
      <div
        class="text-center"
      >
        <h1
          class="display-1 fw-bold"
          data-testid="notif1"
        >
          404
        </h1>
        <p
          class="fs-3"
          data-testid="notif2"
        >
           
          <span
            class="text-danger"
          >
            Opps!
          </span>
           Page not found.
        </p>
        <p
          class="lead"
          data-testid="notif3"
        >
          The page you’re looking for doesn’t exist.
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
